index.js

```diff
@@ -73,32 +73,30 @@ class USHINBase {
       );
     }
 
-    const allPoints = new Set();
-
-    if (main) allPoints.add(main);
-
-    for (const shape of Object.keys(shapes)) {
-      const pointIds = shapes[shape];
-      for (const pointId of pointIds) {
-        const point = pointStore[pointId];
-        if (!point) {
-          const error = new Error("Point ID not found in store");
-          error.pointId = pointId;
-          throw error;
-        }
-        if (!point._id) throw new Error("Must specify point ID");
-        if (!point._rev) {
-          await this.addPoint({ createdAt: createdAtTime, ...point });
-        }
-        allPoints.add(pointId);
-        if (point.referenceHistory) {
-          for (const { pointId: referencePoint } of point.referenceHistory) {
-            allPoints.add(referencePoint);
-          }
-        }
+    const allPoints = new Set([main, ...Object.values(shapes).flat()]);
+
+    for (const pointId of allPoints) {
+      const point = pointStore[pointId];
+      if (!point) {
+        const error = new Error("Point ID not found in store");
+        error.pointId = pointId;
+        throw error;
+      }
+      if (!point._id) throw new Error("Must specify point ID");
+      if (!point._rev) {
+        await this.addPoint({ createdAt: createdAtTime, ...point });
       }
     }
 
+    // Convert set to array to avoid iterating over pointIds which are
+    // added inside the conditional block below
+    [...allPoints].forEach((pointId) => {
```

Comment:
Would it make sense to put this object into the loop on line 76 since we're iterating over the points there anyway?

Reply:
That's a good idea! As long as we iterate over something like `[...allPoints]` instead of `allPoints` itself. Because if we add the referenceHistory pointIds to allPoints inside of a for loop which iterates through allPoints, then the referenceHistory pointIds will get looped over as well (which shouldn't happen, since the referenceHistory points don't exist in the pointStore nor do we want to call `this.addPoint` on them).